### PR TITLE
Add a way to specify additional classes by configuration (instead of hardcoding it and recompiling)

### PR DESCRIPTION
**Pull request by batmat**
(PR filed in the context of [JENKINS-55704](https://issues.jenkins-ci.org/browse/JENKINS-55704))

This PR introduces args4j in the codebase to allow using the tool without having to recompile it each time.

The core goal here was to scan JAXB classes, and we decided to make this reusable instead of just hardcoding our need, given we may need soon to do this again for JAXB or other APIs.

* [x] Add options to run the tool without code modification/hardcoding :heavy_check_mark: 

The new way of using the tool will look like the following (renders as https://gistpreview.github.io/?8016715184ded96b93f631b11c64f972/demo-usage-by-plugin-only-additionalClasses.html):

```shell
# Build it if needed
$ mvn clean package
# Run it (help mode to show existing options)
$ java -jar target/deprecated-usage-in-plugins-0.1-SNAPSHOT-jar-with-dependencies.jar --help
 --additionalClasses FILENAME  : File name for additional classes to scan
 --onlyAdditionalClasses       : Only include in the report the specified
                                 classes (default: false)
 --onlyIncludeJenkinsClasses   : Only include in the report Jenkins related
                                 classes (jenkins.*, hudson.*, etc. (default:
                                 false)
 -c (--includeJavaCoreClasses) : Include classes from java.* and javax.* in the
                                 report (not included by default) (default:
                                 false)
 -h (--help)                   : Shows help (default: true)

$ cat add.txt
javax/xml/bind/DatatypeConverterImpl
javax/xml/bind/DatatypeConverter
$ java -jar target/deprecated-usage-in-plugins-0.1-SNAPSHOT-jar-with-dependencies.jar --additionalClasses add.txt --onlyAdditionalClasses
Downloaded update-center.json
All files are up to date (1590 plugins)
Analyzing deprecated api in Jenkins
add.txt found, adding 2 classes
        adding javax/xml/bind/DatatypeConverterImpl
        adding javax/xml/bind/DatatypeConverter
Analyzing deprecated usage in plugins
.......... .......... .......... .......... ..........
.......... .......... .......... .......... ..........
.......... .......... .......... .......... ..........
.........Writing output/usage-by-plugin.json
Writing output/usage-by-plugin.html
Writing output/deprecated-and-unused.json
Writing output/deprecated-and-unused.html
Writing output/usage-by-api.json
Writing output/usage-by-api.html
duration : 7150 ms at 8 févr. 2019 12:17:45
```

@jenkins-infra/java11-support and @daniel-beck who's used this tool regularly.

### CODE REVIEW

**Comment by batmat** [comment]
I consider this mostly ready for review. There's a last small misbehavior, but nothing blocking.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+FROM maven:3.6.0-jdk-11
+LABEL maintainer="Baptiste Mathus <batmat@batmat.net>"
+
+RUN mkdir /project
+VOLUME /project/work
+ADD . /project
+WORKDIR /project
+RUN mvn clean package && \
+    cd target && \
+    mv deprecated-usage-in-plugins-*-SNAPSHOT-jar-with-dependencies.jar deprecated-usage-fat.jar
```
Comment (alecharp) [inline Dockerfile]: that is one of the reason on change the `<finalName/>` in Maven

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+ENTRYPOINT ["java", "-jar", "target/deprecated-usage-fat.jar"]
```
Comment (alecharp) [inline Dockerfile]: couldn't we use a multi-stage Dockerfile so that to run the jar we don't need the Maven?

File: pom.xml [hunk 3]
```diff
@@ -77,8 +82,8 @@
 				<artifactId>maven-compiler-plugin</artifactId>
 				<version>3.3</version>
 				<configuration>
-					<source>1.7</source>
-					<target>1.7</target>
+					<source>1.8</source>
```
Comment (alecharp) [inline pom.xml]: also why I tend to define `<maven.compiler.target/>` and `<maven.compiler.source/>` in properties to not redefine the complete `build >> plugins >> plugin >> configuration`.

File: Dockerfile (hunk 2 above)
Comment (alecharp) [reply Dockerfile]: Just read the comment on b6a7147 and it's fine with me.

File: Dockerfile (hunk 1 above)
Comment (batmat) [reply Dockerfile]: I didn't check, but I suspect it's not usable anyway here, this is why I simply relied on Docker while building. AFAIK, `finalName` will apply to the usual .jar being generated, not the `...-jar-with-dependencies.jar` one.

File: pom.xml (hunk 3 above)
Comment (batmat) [reply pom.xml]: Yeah, I think there are still many things improvable like these indeed. I am happy to make another pass once this is merged to see where we can make things more current/clean. For instance, I'd like to remove all the currently commented out code (like the `// log...` calls + the various `System.out.println` calls, which should be either removed, or made to be displayed by adding e.g. a new `-verbose` mode in the `Options` class).

**Comment by batmat** [comment]
Side note: I'm aware we might want to (should?) likely rename a few classes/packages to something along `usage` or `code_usage` instead of `deprecatedusage` given this tool is going to be able in standard to do more than just looking for code marked `@Deprecated` with this addition.

File: Dockerfile (hunk 1 above)
Comment (alecharp) [reply Dockerfile]: ok, I went through the Maven Assembly plugin docs again and you are correct. The `<finalName/>` is not used anymore. It used to be when the plugin had the goal `assembly:assembly`.

**Comment by varyvol** [comment]
Looks good in general. My only doubt is what would happen if someone calls this new version in the old way, without providing options and so on? Everything would keep working as it previously did?

**Comment by batmat** [comment]
@varyvol I believe the only place this is really called is the CI really. So, you're quite right I need to check this still works with the previous way (`mvn clean package exec:java`), or adapt the `Jenkinsfile`. But I _think_ this still behaves the same if no CLI switch at all is passed.

Apart from that automated use, I believe that 1) anyway there are only a handful of users of this tool out there and 2) adapting the usage to the new way with switches for interactive usage.

**Comment by varyvol** [comment]
@batmat if you feel confident about it, I'm fine.

File: src/main/java/org/jenkinsci/deprecatedusage/DeprecatedUsage.java [hunk 4]
```diff
@@ -125,7 +107,42 @@ void methodCalled(String className, String name, String desc) {
...
+                Options.getAdditionalClasses().stream().anyMatch(className::startsWith)) {
+            return true;
+        }
+
+        if(Options.get().onlyAdditionalClasses) {
+            return false;
         }
+
+        // Calls to java and javax are ignored by default if not explicitly requested
+        if(className.startsWith("java/") || className.startsWith("javax")) {
```
Comment (MRamonLeon) [inline src/main/java/org/jenkinsci/deprecatedusage/DeprecatedUsage.java]: Isn't it `isJavaClass`?

File: USAGE.adoc [hunk 5]
```diff
@@ -0,0 +1,52 @@
+= How to use this tool
```
Comment (MRamonLeon) [inline USAGE.adoc]: Can you also change the README.md to indicate the purpose of this tool. IIUC it has changed with your improvements, now it does more than find deprecated API. Actually, it should deserve a change on its name, but as it's a tool I don't think is worth it.

File: src/main/java/org/jenkinsci/deprecatedusage/DeprecatedUsage.java (hunk 4 above)
Comment (batmat) [reply src/main/java/org/jenkinsci/deprecatedusage/DeprecatedUsage.java]: Good catch, addressed in https://github.com/jenkins-infra/deprecated-usage-in-plugins/pull/11/commits/13e0045a35ae74f0bb380ec7280a7824026189d1

File: USAGE.adoc (hunk 5 above)
Comment (batmat) [reply USAGE.adoc]: Facelift applied to the README. 

Likely will convert it to asciidoc later, but that's OK for now.

**Pull request merged**